plugins/rh-rbac-backend/src/service/policy-builder.ts

```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
+      });
+
+      if (decision.result === AuthorizeResult.DENY) {
+        throw new NotAllowedError(); // 403
+      }
+
+      const policyRaw: EntityReferencedPolicy = request.body;
+      const err = validatePolicy(policyRaw);
+      if (err) {
+        throw new InputError( // 400
+          `Invalid policy definition. Cause: ${err.message}`,
+        );
+      }
+
+      const policy = transformPolicyToArray(policyRaw);
+
+      if (await enforcer.hasPolicy(...policy)) {
+        throw new ConflictError(); // 409
+      }
+
+      const isAdded = await enforcer.addPolicy(...policy);
+      if (!isAdded) {
+        throw new ServiceUnavailableError(); // 500
+      }
+      response.status(201).end();
+    });
+
+    router.put('/policy/:namespace/:id', async (req, resp) => {
+      const decision = await authorize(env.identity, req, permissions, {
+        permission: policyEntityUpdatePermission,
+      });
+
+      if (decision.result === AuthorizeResult.DENY) {
+        throw new NotAllowedError(); // 403
+      }
+
+      const entityRef = getEntityReference(req);
+
+      const oldPolicyRaw = req.body.oldPolicy;
+      if (!oldPolicyRaw) {
+        throw new InputError(`'oldPolicy' object must be present`); // 400
+      }
+      const newPolicyRaw = req.body.newPolicy;
+      if (!newPolicyRaw) {
+        throw new InputError(`'newPolicy' object must be present`); // 400
+      }
+
+      oldPolicyRaw.entityReference = entityRef;
+      let err = validatePolicy(oldPolicyRaw);
+      if (err) {
+        throw new InputError( // 400
+          `Invalid old policy object. Cause: ${err.message}`,
+        );
+      }
+      newPolicyRaw.entityReference = entityRef;
+      err = validatePolicy(newPolicyRaw);
+      if (err) {
+        throw new InputError( // 400
+          `Invalid new policy object. Cause: ${err.message}`,
+        );
+      }
+
+      const oldPolicy = transformPolicyToArray(oldPolicyRaw);
+      const newPolicy = transformPolicyToArray(newPolicyRaw);
+
+      if (await enforcer.hasPolicy(...newPolicy)) {
+        if (isEqual(oldPolicy, newPolicy)) {
+          resp.status(204).end();
+          return;
+        }
+        throw new ConflictError(); // 409
+      }
+
+      if (!(await enforcer.hasPolicy(...oldPolicy))) {
+        throw new NotFoundError(); // 404
+      }
+
+      // enforcer.updatePolicy(oldPolicyPermission, newPolicyPermission) was not implemented
+      // for ORMTypeAdapter.
+      // So, let's compensate this combination delete + create.
+      const isRemoved = await enforcer.removePolicy(...oldPolicy);
+      if (!isRemoved) {
+        throw new ServiceUnavailableError(); // 500
+      }
+
+      const isAdded = await enforcer.addPolicy(...newPolicy);
+      if (!isAdded) {
+        throw new ServiceUnavailableError(); // 500
+      }
+
+      resp.status(200).end();
     });
 
     return router;
   }
 }
+
+function getEntityReference(req: Request): string {
```

Comment:
Are there unit tests for these utility methods?

Reply:
I think it make sense to align our api urls with catalog entity service: https://github.com/backstage/backstage/blob/41a2601ef613ad90e1900566bbbef80646e73e5c/plugins/catalog-backend/src/service/createRouter.ts#L190
Regarding unit tests, I believe we have adequately covered the invocation of this method by utilizing various URLs in our REST API tests.